Allow client application to control when transaction broadcast protocols are restarted

philipr-za opened the pull request:
## Description
Previously the wallet restarted its transaction broadcast and negotiation protocols as soon as a base node pubkey was set. This caused problems in the base node where on a restart the mempool was empty but the chain had yet to sync so validly mined transactions were cancelled because their chain monitoring protocol started before the blocks were synced.

This PR separates the functions to restart the transaction protocols from the set_base_node_public_key function in the transaction service. This allows the transaction negotiation protocols to be started ASAP as they don’t need a base node and then the client application can choose when to start the broadcast protocols.

The base node application was updated to wait for the local node to be synced before starting the transaction broadcast protocols. The FFI library was updated to start these protocol right after the base node public key was supplied to retain the previous behaviour in the mobile wallets.

## How Has This Been Tested?
tests updated

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
* [ ] Bug fix (non-breaking change which fixes an issue)
* [x] New feature (non-breaking change which adds functionality)
* [ ] Breaking change (fix or feature that would cause existing functionality to change)
* [ ] Feature refactor (No new feature or functional changes, but performance or technical debt improvements)
* [ ] New Tests
* [ ] Documentation

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
* [x] I'm merging against the `development` branch.
* [x] I ran `cargo-fmt --all` before pushing.
* [ ] I have squashed my commits into a single commit.
* [ ] My change requires a change to the documentation.
* [ ] I have updated the documentation accordingly.
* [x] I have added tests to cover my changes.

hansieodendaal commented on applications/tari_base_node/src/builder.rs:
```diff
@@ -467,12 +467,45 @@ where
         .set_base_node_public_key(base_node_public_key.clone())
         .await
         .expect("Problem setting local base node public key for transaction service.");
+    transaction_service_handle
```
I think this should also be delayed until the initial sync has been confirmed, which should be the main priority at startup. For example, **has the Tx been mined** is dependent on an up to date blockchain.

philipr-za replied:
So with this refactor all this does is set the base node pub key, the protocols are only started when the explicit functions are called which is done in the task that waits for the sync to complete. The service itself needs to be started so that its handles are available to the other services.

hansieodendaal replied:
Please see comment below

hansieodendaal reviewed (commented):
Nice refactoring! This logic makes a lot of sense.

Just one comment on when to start the transaction service for a base node wallet.

hansieodendaal commented on applications/tari_base_node/src/builder.rs:
```diff
@@ -467,12 +467,45 @@ where
         .await
         .expect("Problem setting local base node public key for transaction service.");
+    transaction_service_handle
+        .restart_transaction_protocols()
```
This is the actual context; it looks like the transaction protocols are started before initial sync has been achieved, unlike `.restart_broadcast_protocols()`

philipr-za replied:
Transaction protocols are the negotiation protocols with other wallets so this means that the wallet is immediately available to receive/complete negotiation of more transactions, any broadcast and chain monitoring protocols that start from scratch before the BN is synced are fine, its the restart of those protocols before the bn is synced that is the problem.

philipr-za replied:
Ok point taken, to avoid any potential issues in the BN wallet I have made this change.

hansieodendaal commented:
LGTM

philipr-za commented:
needs to be updated in light of https://github.com/tari-project/tari/pull/2239

philipr-za committed:
Allow client application to control when transaction broadcast protocols are restarted

Previously the wallet restarted its transaction broadcast and negotiation protocols as soon as a base node pubkey was set. This caused problems in the base node where on a restart the mempool was empty but the chain had yet to sync so validly mined transactions were cancelled because their chain monitoring protocol started before the blocks were synced.

This PR separates the functions to restart the transaction protocols from the set_base_node_public_key function in the transaction service. This allows the transaction negotiation protocols to be started ASAP as they don’t need a base node and then the client application can choose when to start the broadcast protocols.

The base node application was updated to wait for the local node to be synced before starting the transaction broadcast protocols. The FFI library was updated to start these protocol right after the base node public key was supplied to retain the previous behaviour in the mobile wallets.
